File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/viewmodel/SettingsViewModel.kt
package io.github.notkanaa.equipe.core.viewmodel

import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.InputValidation
import io.github.notkanaa.equipe.core.NotificationAuthorization
import io.github.notkanaa.equipe.core.UserProfile
import io.github.notkanaa.equipe.core.logic.ReminderLeadTime
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlin.coroutines.cancellation.CancellationException

// Port of TeamTasksCore/ViewModels/Settings/SettingsViewModel.swift.

/** State of the « Réglages » tab. */
data class SettingsState(
    /** E-mail of the account. */
    val email: String?,
    /** Reminder lead time (persisted). */
    val leadTime: ReminderLeadTime,
    val profile: UserProfile? = null,
    /** Editable display name (filled by the first load). */
    val displayName: String = "",
    val displayNameError: String? = null,
    val isSavingName: Boolean = false,
    /** The ntfy topic when push is enabled. */
    val pushTopic: String? = null,
    val isUpdatingPush: Boolean = false,
    /** null until loaded. */
    val notificationStatus: NotificationAuthorization? = null,
    /** What the user typed in the deletion confirmation field. */
    val deleteConfirmation: String = "",
    val isSigningOut: Boolean = false,
    val isDeletingAccount: Boolean = false,
    val loadState: LoadState = LoadState.Idle,
    override val error: ErrorState? = null,
) : ErrorHolder {
    val canSaveDisplayName: Boolean
        get() {
            val trimmed = InputValidation.trimmed(displayName)
            return trimmed.isNotEmpty() && trimmed != profile?.displayName && !isSavingName
        }

    val leadTimeOptions: List<ReminderLeadTime> get() = ReminderLeadTime.entries.toList()

    val isPushEnabled: Boolean get() = pushTopic != null

    /** Web address of the topic (also what the ntfy app subscribes to), e.g. `https://ntfy.sh/equipe-…`. */
    val pushTopicUrl: String? get() = pushTopic?.let { "${SettingsViewModel.NTFY_SERVER}/$it" }

    /** Opens (subscribes to) the topic in the ntfy app when installed: `ntfy://ntfy.sh/<topic>`. */
    val ntfyAppUrl: String? get() = pushTopic?.let { "ntfy://ntfy.sh/$it" }

    /** « Activées », « Refusées », « Pas encore demandées », « … » (not loaded). */
    val notificationStatusText: String
        get() = when (notificationStatus) {
            NotificationAuthorization.AUTHORIZED -> "Activées"
            NotificationAuthorization.DENIED -> "Refusées"
            NotificationAuthorization.NOT_DETERMINED -> "Pas encore demandées"
            null -> "…"
        }

    /** Shown when refused (the permission can only be changed in the system settings) or not yet asked. */
    val notificationHint: String?
        get() = when (notificationStatus) {
            NotificationAuthorization.DENIED ->
                "Pour recevoir les rappels et les nouvelles tâches, autorisez les notifications d’Équipe dans les " +
                    "Paramètres du téléphone."
            NotificationAuthorization.NOT_DETERMINED ->
                "Autorisez les notifications pour recevoir les rappels d’échéance et les nouvelles tâches."
            NotificationAuthorization.AUTHORIZED, null -> null
        }

    val canRequestNotifications: Boolean get() = notificationStatus == NotificationAuthorization.NOT_DETERMINED

    /** True once « SUPPRIMER » is typed exactly. */
    val canDeleteAccount: Boolean
        get() = deleteConfirmation == SettingsViewModel.DELETE_CONFIRMATION_WORD && !isDeletingAccount
}

/**
 * « Réglages » tab: display name, reminder lead time, optional ntfy push, notification permission, sign-out and account
 * deletion (only after typing « SUPPRIMER »).
 *
 * Screen: `LaunchedEffect(model) { model.autoRefresh() }`. Sign-out and deletion end the session through
 * `AuthService.authStates()`: `AppModel` then shows the login screen.
 */
class SettingsViewModel(
    val session: SessionModel,
    scope: CoroutineScope,
) : ScreenModel<SettingsState>(
    SettingsState(email = session.user.email, leadTime = ReminderLeadTime.load(session.platform.store)),
    { state, error -> state.copy(error = error) },
) {
    private val runner = LoadRunner(scope)
    private val background = BackgroundWork(scope)
    private var loadedRevision: Int? = null
    private var fetchingRevision: Int? = null

    /** Editable display name. */
    var displayName: String
        get() = current.displayName
        set(value) = update {
            it.copy(
                displayName = value,
                displayNameError = if (it.displayNameError != null) SignUpViewModel.displayNameMessage(value) else null,
            )
        }

    /** What the user typed in the deletion confirmation field. */
    var deleteConfirmation: String
        get() = current.deleteConfirmation
        set(value) = update { it.copy(deleteConfirmation = value) }

    /** Picker binding: persists the choice and resynchronizes the reminders in the background. */
    var leadTime: ReminderLeadTime
        get() = current.leadTime
        set(value) {
            if (value == current.leadTime) return
            store(value)
            background.start { session.synchronizeReminders() }
        }

    // region Loading

    /** Changes when the profile may have changed on another device (a display-name update is a memberships signal). */
    val refreshKey: RefreshKey get() = RefreshKey(session.feed.membershipsRevision.value)

    /** [refreshKey] as a flow: the current value first, then every change. */
    val refreshKeys: Flow<RefreshKey> = session.feed.membershipsRevision.map { RefreshKey(it) }.distinctUntilChanged()

    val needsRefresh: Boolean
        get() = current.loadState != LoadState.Loaded || loadedRevision != refreshKey.revision

    suspend fun load() {
        if (!needsRefresh) return
        val upToDate = runner.isRunning && fetchingRevision == refreshKey.revision
        runner.run(rerunIfRunning = !upToDate) { fetch() }
    }

    suspend fun reload() {
        runner.run(rerunIfRunning = true) { fetch() }
    }

    /** Loads now, then again whenever [refreshKey] changes. Never returns: run it in a `LaunchedEffect`. */
    suspend fun autoRefresh() {
        refreshKeys.collectLatest { load() }
    }

    private suspend fun fetch() {
        val revision = refreshKey.revision
        fetchingRevision = revision
        try {
            update { it.copy(loadState = startedLoad(it.loadState)) }
            val status = session.platform.notifications.authorizationStatus()
            update { it.copy(notificationStatus = status) }
            val profiles = session.services.profiles
            val push = session.services.push
            val (profile, topic) = coroutineScope {
                val profileRequest = async { profiles.myProfile() }
                val topicRequest = async { push.currentTopic() }
                profileRequest.await() to topicRequest.await()
            }
            loadedRevision = revision
            update {
                // Keep a name being edited; follow the server otherwise.
                val keepEdited = it.profile != null && it.displayName != it.profile.displayName
                it.copy(
                    displayName = if (keepEdited) it.displayName else profile.displayName,
                    profile = profile,
                    pushTopic = topic,
                    email = session.user.email,
                    loadState = LoadState.Loaded,
                )
            }
        } catch (error: CancellationException) {
            update { it.copy(loadState = cancelledLoad(it.loadState)) }
            throw error
        } catch (error: Exception) {
            update {
                val failure = loadFailure(it.loadState, it.error, error)
                it.copy(loadState = failure.loadState, error = failure.error)
            }
        } finally {
            fetchingRevision = null
        }
    }

    // endregion

    // region Display name

    suspend fun saveDisplayName(): Boolean {
        if (current.isSavingName) return false
        dismissError()
        val message = SignUpViewModel.displayNameMessage(current.displayName)
        update { it.copy(displayNameError = message) }
        if (message != null) return false
        update { it.copy(isSavingName = true) }
        try {
            val updated = session.services.profiles.updateDisplayName(current.displayName)
            update { it.copy(profile = updated, displayName = updated.displayName) }
            // Names appear on every group screen.
            session.feed.bumpAll()
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            if (present(error) == AppError.InvalidDisplayName) {
                update { it.copy(displayNameError = AppError.InvalidDisplayName.messageFR, error = null) }
            }
            return false
        } finally {
            update { it.copy(isSavingName = false) }
        }
    }

    // endregion

    // region Reminders

    /**
     * Persists the lead time and resynchronizes the reminders (awaited). Returns false when « Mes tâches » could not be
     * loaded (the reminders then keep their previous schedule until the next synchronization).
     */
    suspend fun setLeadTime(value: ReminderLeadTime): Boolean {
        store(value)
        return session.synchronizeReminders()
    }

    private fun store(value: ReminderLeadTime) {
        update { it.copy(leadTime = value) }
        value.save(session.platform.store)
    }

    /** Waits for the background work started by the [leadTime] setter (tests). */
    internal suspend fun waitForBackgroundWork() {
        background.waitForAll()
    }

    // endregion

    // region Push (ntfy)

    suspend fun enablePush(): Boolean {
        if (current.isUpdatingPush) return false
        update { it.copy(error = null, isUpdatingPush = true) }
        try {
            val topic = session.services.push.enable()
            update { it.copy(pushTopic = topic) }
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            present(error)
            return false
        } finally {
            update { it.copy(isUpdatingPush = false) }
        }
    }

    suspend fun disablePush(): Boolean {
        if (current.isUpdatingPush) return false
        update { it.copy(error = null, isUpdatingPush = true) }
        try {
            session.services.push.disable()
            update { it.copy(pushTopic = null) }
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            present(error)
            return false
        } finally {
            update { it.copy(isUpdatingPush = false) }
        }
    }

    /** Switch helper. */
    suspend fun setPushEnabled(enabled: Boolean): Boolean = if (enabled) enablePush() else disablePush()

    // endregion

    // region Notification permission

    suspend fun refreshNotificationStatus() {
        val status = session.platform.notifications.authorizationStatus()
        update { it.copy(notificationStatus = status) }
    }

    /** Asks the system for the permission (only possible once); when granted, reminders and catch-up run. */
    suspend fun requestNotifications(): Boolean {
        val status = session.requestNotificationAuthorizationIfNeeded()
        update { it.copy(notificationStatus = status) }
        return status == NotificationAuthorization.AUTHORIZED
    }

    // endregion

    // region Session

    suspend fun signOut(): Boolean {
        if (current.isSigningOut) return false
        update { it.copy(error = null, isSigningOut = true) }
        try {
            session.services.auth.signOut()
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            present(error)
            return false
        } finally {
            update { it.copy(isSigningOut = false) }
        }
    }

    /** Deletes the account (only when « SUPPRIMER » was typed exactly). */
    suspend fun deleteAccount(): Boolean {
        if (current.deleteConfirmation != DELETE_CONFIRMATION_WORD) {
            present(DELETE_CONFIRMATION_REQUIRED_MESSAGE, AppError.InvalidInput)
            return false
        }
        if (current.isDeletingAccount) return false
        update { it.copy(error = null, isDeletingAccount = true) }
        try {
            session.services.auth.deleteAccount()
            session.platform.store.set(MyTasksViewModel.lastSeenKey(session.userId), null)
            update { it.copy(deleteConfirmation = "") }
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            present(error)
            return false
        } finally {
            update { it.copy(isDeletingAccount = false) }
        }
    }

    // endregion

    companion object {
        const val DELETE_CONFIRMATION_WORD: String = "SUPPRIMER"

        const val DELETE_CONFIRMATION_REQUIRED_MESSAGE: String = "Tapez « SUPPRIMER » pour confirmer."

        const val DELETE_ACCOUNT_WARNING: String =
            "Votre compte, votre profil et vos assignations seront supprimés définitivement. Les groupes dont vous " +
                "êtes le seul membre seront supprimés avec leurs tâches ; dans les autres, le membre le plus ancien " +
                "deviendra admin si vous étiez le seul admin. Tapez « SUPPRIMER » pour confirmer."

        const val NTFY_SERVER: String = "https://ntfy.sh"

        /** Google Play page of the free ntfy app (Android addition). */
        const val NTFY_PLAY_STORE_URL: String = "https://play.google.com/store/apps/details?id=io.heckel.ntfy"

        /** Steps shown when push is enabled (the first one names the Android store instead of the App Store). */
        val PUSH_INSTRUCTION_STEPS: List<String> = listOf(
            "Installez l’application gratuite « ntfy » depuis le Play Store (ou F-Droid).",
            "Dans ntfy, touchez « + » puis abonnez-vous au sujet ci-dessous (serveur ntfy.sh).",
            "Une notification vous prévient quand une tâche vous est assignée, même quand Équipe est fermée.",
        )

        const val PUSH_PRIVACY_NOTE: String =
            "Gardez ce sujet secret : toute personne qui le connaît peut voir quand une tâche vous est assignée " +
                "(le titre de la tâche n’est jamais envoyé)."

        const val PUSH_DISABLED_EXPLANATION: String =
            "Recevez une notification quand une tâche vous est assignée, même quand Équipe est fermée, grâce à " +
                "l’application gratuite ntfy."
    }
}

File: Packages/TeamTasksKit/Sources/TeamTasksCore/ViewModels/Settings/SettingsViewModel.swift
import Foundation
import Observation

/// « Réglages » tab: display name, reminder lead time, optional ntfy push, notification permission, sign-out and
/// account deletion (only after typing « SUPPRIMER »).
///
/// View: `.task(id: model.refreshKey) { await model.load() }`. Sign-out and deletion end the session through
/// `AuthService.authStates()`: `AppModel` then shows the login screen.
@MainActor
@Observable
public final class SettingsViewModel: ErrorPresenting {
    public static let deleteConfirmationWord = "SUPPRIMER"
    public static let deleteAccountWarning =
        "Votre compte, votre profil et vos assignations seront supprimés définitivement. Les groupes dont vous êtes le seul membre seront supprimés avec leurs tâches\u{00A0}; dans les autres, le membre le plus ancien deviendra admin si vous étiez le seul admin. Tapez «\u{00A0}SUPPRIMER\u{00A0}» pour confirmer."
    public static let ntfyServer = "https://ntfy.sh"
    public static let pushInstructionSteps = [
        "Installez l’application gratuite «\u{00A0}ntfy\u{00A0}» depuis l’App Store.",
        "Dans ntfy, touchez «\u{00A0}+\u{00A0}» puis abonnez-vous au sujet ci-dessous (serveur ntfy.sh).",
        "Une notification vous prévient quand une tâche vous est assignée, même quand Équipe est fermée.",
    ]
    public static let pushPrivacyNote =
        "Gardez ce sujet secret\u{00A0}: toute personne qui le connaît peut voir quand une tâche vous est assignée (le titre de la tâche n’est jamais envoyé)."
    public static let pushDisabledExplanation =
        "Recevez une notification quand une tâche vous est assignée, même quand Équipe est fermée, grâce à l’application gratuite ntfy."

    // MARK: Profile

    public private(set) var profile: UserProfile?
    /// Editable display name (filled by `load()`).
    public var displayName: String {
        get { displayNameValue }
        set {
            displayNameValue = newValue
            if displayNameError != nil { displayNameError = SignUpViewModel.displayNameMessage(newValue) }
        }
    }

    public private(set) var displayNameError: String?
    public private(set) var isSavingName = false

    // MARK: Push

    public private(set) var pushTopic: String?
    public private(set) var isUpdatingPush = false

    // MARK: Notifications

    /// nil until loaded.
    public private(set) var notificationStatus: NotificationAuthorization?

    // MARK: Account

    /// What the user typed in the deletion confirmation field.
    public var deleteConfirmation = ""
    public private(set) var isSigningOut = false
    public private(set) var isDeletingAccount = false

    public private(set) var loadState: LoadState = .idle
    public var error: ErrorState?

    public let session: SessionModel
    private var displayNameValue = ""
    private var leadTimeValue: ReminderLeadTime?
    private let runner = LoadRunner()
    private let background = BackgroundWork()
    private var loadedRevision: Int?
    private var fetchingRevision: Int?

    public init(session: SessionModel) {
        self.session = session
    }

    // MARK: - Loading

    /// Changes when the profile may have changed on another device (a display-name update is a memberships signal).
    public var refreshKey: RefreshKey { RefreshKey(revision: session.feed.membershipsRevision) }

    public var needsRefresh: Bool { loadState != .loaded || loadedRevision != refreshKey.revision }

    public func load() async {
        guard needsRefresh else { return }
        let upToDate = runner.isRunning && fetchingRevision == refreshKey.revision
        await runner.run(rerunIfRunning: !upToDate) { [weak self] in await self?.fetch() }
    }

    public func reload() async {
        await runner.run(rerunIfRunning: true) { [weak self] in await self?.fetch() }
    }

    private func fetch() async {
        let revision = refreshKey.revision
        fetchingRevision = revision
        defer { fetchingRevision = nil }
        if loadState != .loaded { loadState = .loading }
        if leadTimeValue == nil {
            leadTimeValue = ReminderLeadTime.load(from: session.platform.store)
        }
        notificationStatus = await session.platform.notifications.authorizationStatus()
        let profiles = session.services.profiles
        let push = session.services.push
        do {
            async let profileRequest = profiles.myProfile()
            async let topicRequest = push.currentTopic()
            let (profile, topic) = try await (profileRequest, topicRequest)
            // Keep a name being edited; follow the server otherwise.
            if self.profile == nil || displayNameValue == self.profile?.displayName {
                displayNameValue = profile.displayName
            }
            self.profile = profile
            pushTopic = topic
            loadedRevision = revision
            loadState = .loaded
        } catch {
            guard let state = ErrorState(from: error) else {
                if loadState == .loading { loadState = .idle }
                return
            }
            if loadState == .loaded {
                self.error = state
            } else {
                loadState = .failed(state.message)
            }
        }
    }

    // MARK: - Account info

    /// E-mail of the account.
    public var email: String? { session.user.email }

    // MARK: - Display name

    public var canSaveDisplayName: Bool {
        let trimmed = InputValidation.trimmed(displayNameValue)
        return !trimmed.isEmpty && trimmed != profile?.displayName && !isSavingName
    }

    @discardableResult
    public func saveDisplayName() async -> Bool {
        guard !isSavingName else { return false }
        error = nil
        displayNameError = SignUpViewModel.displayNameMessage(displayNameValue)
        guard displayNameError == nil else { return false }
        isSavingName = true
        defer { isSavingName = false }
        do {
            let updated = try await session.services.profiles.updateDisplayName(displayNameValue)
            profile = updated
            displayNameValue = updated.displayName
            // Names appear on every group screen.
            session.feed.bumpAll()
            return true
        } catch {
            if present(error) == .invalidDisplayName {
                displayNameError = AppError.invalidDisplayName.messageFR
                self.error = nil
            }
            return false
        }
    }

    // MARK: - Reminders

    public var leadTimeOptions: [ReminderLeadTime] { ReminderLeadTime.allCases }

    /// Picker binding: persists the choice and resynchronizes the reminders in the background.
    public var leadTime: ReminderLeadTime {
        get { leadTimeValue ?? ReminderLeadTime.load(from: session.platform.store) }
        set {
            guard newValue != leadTime else { return }
            store(newValue)
            background.start { [weak self] in
                _ = await self?.session.synchronizeReminders()
            }
        }
    }

    /// Persists the lead time and resynchronizes the reminders (awaited). Returns false when « Mes tâches » could
    /// not be loaded (the reminders then keep their previous schedule until the next synchronization).
    @discardableResult
    public func setLeadTime(_ value: ReminderLeadTime) async -> Bool {
        store(value)
        return await session.synchronizeReminders()
    }

    private func store(_ value: ReminderLeadTime) {
        leadTimeValue = value
        value.save(to: session.platform.store)
    }

    /// Waits for the background work started by the `leadTime` setter (tests).
    func waitForBackgroundWork() async {
        await background.waitForAll()
    }

    // MARK: - Push (ntfy)

    public var isPushEnabled: Bool { pushTopic != nil }

    /// Web address of the topic (also what the ntfy app subscribes to).
    public var pushTopicURL: URL? {
        pushTopic.flatMap { URL(string: "\(Self.ntfyServer)/\($0)") }
    }

    /// Opens the topic in the ntfy app when installed.
    public var ntfyAppURL: URL? {
        pushTopic.flatMap { URL(string: "ntfy://ntfy.sh/\($0)") }
    }

    @discardableResult
    public func enablePush() async -> Bool {
        guard !isUpdatingPush else { return false }
        error = nil
        isUpdatingPush = true
        defer { isUpdatingPush = false }
        do {
            pushTopic = try await session.services.push.enable()
            return true
        } catch {
            present(error)
            return false
        }
    }

    @discardableResult
    public func disablePush() async -> Bool {
        guard !isUpdatingPush else { return false }
        error = nil
        isUpdatingPush = true
        defer { isUpdatingPush = false }
        do {
            try await session.services.push.disable()
            pushTopic = nil
            return true
        } catch {
            present(error)
            return false
        }
    }

    /// Toggle binding helper.
    @discardableResult
    public func setPushEnabled(_ enabled: Bool) async -> Bool {
        if enabled {
            return await enablePush()
        }
        return await disablePush()
    }

    // MARK: - Notifications permission

    /// « Activées », « Refusées », « Pas encore demandées ».
    public var notificationStatusText: String {
        switch notificationStatus {
        case .authorized: "Activées"
        case .denied: "Refusées"
        case .notDetermined: "Pas encore demandées"
        case nil: "…"
        }
    }

    /// Shown when refused: the permission can only be changed in the system settings.
    public var notificationHint: String? {
        switch notificationStatus {
        case .denied: "Pour recevoir les rappels et les nouvelles tâches, autorisez les notifications d’Équipe dans l’app Réglages de l’iPhone."
        case .notDetermined: "Autorisez les notifications pour recevoir les rappels d’échéance et les nouvelles tâches."
        case .authorized, nil: nil
        }
    }

    public var canRequestNotifications: Bool { notificationStatus == .notDetermined }

    public func refreshNotificationStatus() async {
        notificationStatus = await session.platform.notifications.authorizationStatus()
    }

    /// Asks the system for the permission (only possible once); when granted, reminders and catch-up run.
    @discardableResult
    public func requestNotifications() async -> Bool {
        let status = await session.requestNotificationAuthorizationIfNeeded()
        notificationStatus = status
        return status == .authorized
    }

    // MARK: - Session

    @discardableResult
    public func signOut() async -> Bool {
        guard !isSigningOut else { return false }
        error = nil
        isSigningOut = true
        defer { isSigningOut = false }
        do {
            try await session.services.auth.signOut()
            return true
        } catch {
            present(error)
            return false
        }
    }

    /// True once « SUPPRIMER » is typed exactly.
    public var canDeleteAccount: Bool { deleteConfirmation == Self.deleteConfirmationWord && !isDeletingAccount }

    /// Deletes the account (only when `canDeleteAccount`).
    @discardableResult
    public func deleteAccount() async -> Bool {
        guard deleteConfirmation == Self.deleteConfirmationWord else {
            present(message: "Tapez «\u{00A0}SUPPRIMER\u{00A0}» pour confirmer.", error: .invalidInput)
            return false
        }
        guard !isDeletingAccount else { return false }
        error = nil
        isDeletingAccount = true
        defer { isDeletingAccount = false }
        do {
            try await session.services.auth.deleteAccount()
            session.platform.store.set(nil, forKey: MyTasksViewModel.lastSeenKey(userId: session.userId))
            deleteConfirmation = ""
            return true
        } catch {
            present(error)
            return false
        }
    }
}
